dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +53,176 @@ public Pattern getParameterPattern() {
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+                String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                allParams.put(NameNormaliser.normaliseName(paramName),
+                              createDbParamAccessor(paramName, rs, getParamDirection(rs)));
             }
-            rs.close();
             return allParams;
         }
+
+        abstract Direction getParamDirection(ResultSet rs) throws SQLException;
+
+        private DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs,
+                                                          Direction paramDirection)
+                                                                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            int position = rs.getInt("ORDINAL_POSITION");
+            return createDbParameterAccessor(paramName, paramDirection,
+                                             typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                                             getJavaClass(paramTypeName), position - 1);
+        }
+    }
+
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(Connection conn, String defaultSchemaName, String objName) {
+            super(conn, defaultSchemaName, objName);
+        }
+
+        @Override
+        public boolean exists() throws SQLException {
+            ResultSet rs = con.getMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+            return rs.next();
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return con.getMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    private class DatabaseProcedure extends DatabaseObject {
+
+        private DatabaseProcedure(Connection conn, String defaultSchemaName, String objName) {
+            super(conn, defaultSchemaName, objName);
+        }
+
+        @Override
+        public boolean exists() throws SQLException {
+            ResultSet rs = con.getMetaData().getProcedures(null, name.getSchemaName(), name.getObjectName());
+            return rs.next();
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return con.getMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs)
+                                        throws SQLException {
+            Direction dir = null;
+            switch (rs.getInt("COLUMN_TYPE")) {
+                case DatabaseMetaData.procedureColumnIn:
+                    dir = Direction.INPUT;
+                    break;
```

Comment:
It can be either return in the default option or it can be return statement after the switch (both compile fine).

But good that you turned attention into that: right now it returns `null` in case of unknown type. Shouldn't it be throwing an exception?